pandas/core/generic.py

```diff
@@ -1912,6 +1912,134 @@ def __array_wrap__(
             self, method="__array_wrap__"
         )
 
+    def __array_ufunc__(
+        self, ufunc: Callable, method: str, *inputs: Any, **kwargs: Any
+    ):
```

Comment:
might consider moving this to pandas/core/arraylike.py or similar, just to remove complexity from generic.py (and then this might be easier to grok).

not a blocker and followon ok, but something to think about.

Reply:
Yeah. It's a lot to throw in here.

I assume you prefer a single implementation on NDFrame rather than Series and DataFrame separately? There's a bunch of shared code (deciding to dispatch / return NotImplemented, aligning, etc) but I did need a couple `self.ndim` checks for series / frame specific behavior.

Reply:
yes exactly, same impl, maybe just pass in ndim itself. ideally just to not have all of this code in generic.

Reply:
could this (eventually) be shared by Index too?

Reply:
Probably. The main difference there will be that there wouldn't be alignment, but that's easy to skip.

Reply:
can we throw this is pandas/core/ufunc.py ? (and just pass in self to evaluate?) alt do it as a mixin to generic.

Reply:
if it can eventually be shared by index/ea then its a good fit for core.arraylike

Reply:
sure ok with that as well (mainly just NOT in generic)